### ISSUE TITLE
Roadmap rough draft

### ISSUE THREAD

**Issue by atanasster** [grommet/grommet#3271]
a grommet roadmap outline to start the conversation, I can expand on the details when i have some more time.

v3 - goal - improve quality and development process.
1. Rewrite components in typescript to use better error finding and also create usable typescript interfaces for users
2. Remove react-desc and adopt a docgen type documentation style using the richer ts interfaces for documentation
3. Update grommet site, generate-readme.js to use typescript prop interfaces
4. Replace grommet cli with rollup for build and tests
5. Monorepo?

v4 - goal : improve testability and performance.
1. Discontinue/discourage use of snapshots as they are limiting the test awareness
2. Use storybooks stories for automated regression tests
3. Integrate profiling checkpoints to optimize current speed and also part of automated tests to catch if a performance regression degradation is introduced

v5 - focus on enterprise
1. Standardize theming and styling (ie styled-design) and potential use of emotion/styled-components
2. DataTable - not really adapted for enterprise apps as it scrolls infinitely instead of the usual paging. Either enhance DataTable of include a third party integration
3. Create grommet integration for all major data table (grid) components - both commercial and freeware. Accompanying article/documentation for each integration.
4. Form validation for all major react form validation tools and articles. Current grommet form validation doesnt look easy to use to me and can get discontinued for preference of third-party integrations.
5. MaskedInput enhancements and examples for all types of data entry
6. discontinue grommet Chart (usage warnings)
7. integration of major charting libraries (use grommet theme colors for chart series, custom tooltip to use grommet components) and accompanying articles 
8. Sample enterprise app ie grommet-ferret++

Next:
1. Testing hooks for third party libraries and commercial tools
2. Reporting library integration and styling
3. Grommet-native for react-native

**Comment by ShimiSun** [comment]
I'd love to hear what more people think about the suggestion, opening it up for the community to respond... please share your feedback!

**Comment by atanasster** [comment]
Following a private conversation
-why typescript rewrite is getting highest priority
The d.ts files were generated initially by react-desc and are full of issues. Although currently those files are maintained manually, it takes almost a full time dev to fix those files. Additionally rewriting with typescript will uncover some bugs and issues with the library itself.

-why replace react-desc with docgen style documentation
All the props information (type, default value, required) can be obtained from the typescript interfaces. All that is left for react-desc is descriptions of properties and components. Docgen style comments can be used by the editors, storybook.js, styleguidist etc. while react-desc is a closed environment that no other tool understands

-why use rollup(or other third party tooling system) for build process
The grommet cli tooling is not well documented, used or maintained making it very hard to have contributors on the build or testing areas

- why testing and performance before enterprise focus
The lack of proper regression tests makes for unreliable release process incluing regression bugs. Before being able to add many new features, its best to stabilize grommet with typescrupt, regression tests

**Comment by oorestisime** [comment]
hi thanks for all these ideas! A few comments

- I think we need an 2.99 version with monorepo and new Storybook. This should lay the way for the the 3.0 version while providing some extra safety measures to avoid breaking changes and bugs (with the visual testing and maybe the jest snapshotying of stories). 
- Some of the 3.0 goals you mention should be unrelated to grommet components such as updating grommet site to use TS. this could also be done now and would produce some great insights for bugs to fix without needing to wait for 3.0.
- I am not sure what you mean with grommet-cli. right now the library is build using webpack and while rollup might be easier or whatever i don't see any reason to switch away from webpack. Its additional work with not much to offer especially not for a 3.0
- I don't agree with disabling or discontinuing jest snapshots, they are powerful and i don't see the reason why they should be removed. Inside these tests we use RTL where we interact with components. they are not just snapshotting empty components. While not perfect or complete they have enabled to catch some bugs in PRs. IMHO we should seek on adding more tests but not removing them as long as they don't produce noise or are useless.
- which brings me to the my point that i feel that naming a v4 without any reason is imho is wrong. automated stories snapshoting and visual testing should be included now (like in a 2.x release) rather than in a v4 and it won't have any impact to the direct user. 
- While i agree we can investigate performance issues i really don't believe the issue is grommet. the performance issues are mainly due to the usage of the context. i reckon with a TS rewrite we will be using internally even more the hooks which should help in that regard. What i am trying to say is that we cannot say we ll focus on a v4 for performance without a clear vision of what we can improve and it can be done in parallel. 

Few suggestions for the 3.0

- explore being compatible with emotion in parallel with styled-components with the babel plugin someone mentioned
-  create a new package called grommet-extras or whatever and include few of the components that have been a bottleneck the last months, bring very little value to the main library, are not used that often, and there are much better alternatives out there: Chart (once you get to want to do something a bit more complicate almost always I end up with chartjs), datatable (same reason not flexible enough), carousel , WorldMap and i would argue even for form. I believe it shouldn't be  grommet's problem to provide these components and maintained in the main library, there are alternatives for each thing and are mostly used once you get to do something a bit more complicated. We (should) provide primitives to enable users to use such external libraries. I.e make Table* good enough so that a user could be their datatable technology and just use these components. </EndLongPoint> This should also help bring down the bundlesize :D

All in all i agree in very much of these  points and direction. I mostly think some things could be done in a different order. 
Once again thanks for the initiative!

**Comment by atanasster** [comment]
Greta, thanks a lot for the conversation. We are pretty much in agreement, so maybe we can start keeping this document somewhere to be easily updated. 

>>I think we need an 2.99 version with monorepo and new Storybook csf format
Sounds good, I have been working with storybooks and can help out a bit.
Monorepo - It has been a common issue with grommet that most of the additional tools/libararies are just dropped after a few months of development, nd this can be an issue if they are now part of a monorepo. So what will the monorepo include besides grommet: 
- the site (seems a must)
- grommet addons
- designer, theme designer - will those stay and be maintained continuously?

>>updating grommet site to use TS
this is related to the typescript conversion as it will need to stop using the react-desc .doc prop tables. Eventually the new Storybook docs page can also display prop tables from docgen comments and could replace the documentation feature on the site.

>>I am not sure what you mean with grommet-cli.
You are right, this was a grommet tool that used to be used to build the project and I dont see it anymore

>>i don't see any reason to switch away from webpack
just rollup is easier to generate many possible output formats from a single (typescript) base, but since the grommet tool is taken out, this point can be removed.

>>I don't agree with disabling or discontinuing jest snapshots, 
I mean the auto-generated snapshots - they are just confusing, noisy and keep providing a false sense of grommet being tested while they are most of the times just being regenerated. Any snapshots that are thoughtfully created, I am fine :)

>>while i agree we can investigate performance issues i really don't believe the issue is grommet. 
Mostly to have performance checkpoints. In the last year+ v2 has stagnated, mostly fixing a flaky initial release. In order to have an active and safe development cycle, I believe grommet needs to sit on a stable set of regression and performance tests. 

>> we cannot say we ll focus on a v4 for performance without a clear vision of what we can improve 
A common misconception in profiling is that the developers probably know where the bottlenecks are. The truth is - without measuring, most of the time optimizing is spent on things that are not the actual bottlenecks 

>>explore being compatible with emotion in parallel with styled-components
agreed, and probably refactor the theme design to be consistant

>>create a new package called grommet-extras or whatever and include few of the components that have been a bottleneck the last months
Chart, WorldMap etc - agreed.

>>We (should) provide primitives to enable users to use such external libraries
Agreed, but also keep in mind that currently the grommet user base is quite small and practically none of the makers of external components has any interest in providing support for grommet. 
That is why I suggest for grommet to at least initially make the integrations (so to also find out what support is needed), write articles with each integration (so the vendors/makers of those external libraries can get some exposure and get at least some interest in grommet), provide any other benefits to anyone interested in growing the grommet ecosystem.

**Comment by oorestisime** [comment]
*monorepo*
I would like to have us publish the patterns in Netlify in each PR and merge into master as another mean for testing. In addition we could have examples folder that builds sites with different build techniques such as rollup, gatsby, next etc. This should keep us ahead of bugs

*snapshots*
There are not automated snapshots, one needs to add them in the tests suite and test something specific. Maybe you are referring to something else but can't make any correlation to existing code.

*performance*
Yeah, what i meant is that we can not plan for a v2 release for improved performance if we don't know there are breaking changes or actual improvements. I believe using profilers and investigating can be in parallel and if we indeed find something to be done in a minor release unless there's a breaking change with it which i really doubt. After all the components are just react components, performance can possibly be found in some logic inside a specific component i can't understand where we can globally improve. 

Lets leave this a few more days open for people to contribute and then I agree we should release the roadmap somehow and create actionable items :)

**Comment by atanasster** [comment]
*snapshots*
I mean `jest --updateSnapshot`- when those generate large amounts of data to compare and are so noisy and/or incomprehensible that they are useless. 

* An example that can be explained easy:
The snapshot for grommet-icons (which should be in the monorepo together with icons-site). The snapshots are just the svgs data, so no human can actually understand if they are correct or not correct. On the other hand, even major bugs can not be caught using those snapshots )let say a user introduced a fixed color in an icon), so most of the time, they are just auto-generated.  

If the snapshots are targeted - lets say you test a specific react prop and verify the outcome on just a few select dom props, then you would not need to autogenerate the snapshots except for possibly on the first writing of the test. 
Of course, the snapshots in grommet could have gone a long way since my last checking them and they might be somewhat useful now.

**Comment by talarico43** [comment]
just want to add a +1 for focus on enterprise. my team is currently building an enterprise application using grommet. the goals outlined in v5 above line up really well with how we have been using grommet thus far.

**Comment by atanasster** [comment]
Thanks @talarico43 great to hear you like the enterprise direction

**Comment by nomadus** [comment]
Regarding the point which mentions grommet-extras package. I don't mind separating some components to a new package, with v1 you already had such a setup. What bothers me is an idea to search for alternatives. Having an enterprise background I could tell that I would prefer stable components even with fewer features from a single vendor. This makes life a lot easier. 

My 2 cents on the overall approach. I might be wrong, but I notice that the companies who develop UI kits tend to create more and more higher-order components, e.g. look at components from Atlassian, Ant, and MS.  I prefer such an approach over nails and hummer approach as this makes my code consistent and myself more effective. Some example of components: TreeTable, Tooltip, Tags, NumberInput (covered by @atanasster), etc.

**Comment by atanasster** [comment]
@nomadus , i wholeheartedly agree that the components that you would use should come into one single-vendor package.
In the case of grommet, all base enterprise type components like cards, editing controls (numberinputs, passwordinputs etc), authentication forms, progress indicators, spinners etc should come into the main package with all the associated documentation, support and constant enhancements.  The v2 mantra ‘grommet provides the building blocks blah blah’ was taken up just because v2 was so poorly released with half the components dropped.

Then add extra higher-level controls like - invoicing, blogging, accounting etc. into separate packages.

The issue is that grommet Charts and even FormValidation, DataTable are not up to par, and honestly i dont see the team making top notch versions of those components. Thus imo a better approach is to drop those (putting them into a grommet-extras package is the same but sounds better) and make integrations with industry-best third party versions of those components.

**Comment by murbanowicz** [comment]
about rewriting in TypeScript - I am here to help .

**Comment by RobertGary1** [comment]
1) Dialog forms that integrate validation on submission. We had to build our own infrastructure around Grommet forms to manage validation
2) More support for higher level components
3) We really need a charting solution. Building charts from primitive pieces isn't really a good solution. Maybe even a tight integration with some other charting library.
4) Infinite scroll from the middle. Scroll up and load more above. This is required when you land in the middle of a table/tile from a link.
5) Working example of end Grommet app to help new people with setting up environment. The examples now are too simple. Need full example of webpack, babel, with support for hot swap dev and packing for distro. 
6) I miss RoutedAnchor
7) Direct support for a variety of notifications (slide outs, etc). We had toast in v1, and there are examples of solutions in v2 but would be better to be a component again.
8) Bring back tiles

**Comment by juanlanus** [comment]
**With focus on documentation only:**  (sent to @britt6612 minutes ago)
I'm thinking of the advantage of defining a set of "common props" (similar to CSS commom attributes) to be (better) defined in a doc by itself to be referenced from the individual control docs.
This would render a two-fold advantage: allowing to focus more effort in defining each common prop while shortening the controls' own docs. Additionally it might add consistence (like in "fill" vs "full").  
IMO the docs are of paramount importance in order to foster adoption.  
And will be a hard requirement for "v5" enterprise buy-in.

**Comment by atanasster** [comment]
Looking at the many great suggestions here, i would like to add an organizational proposal that would prevent the issues re v2

- make a governing body for grommet that explicitly prohibits 2 or more members from the same organization to participate. Obviously i have in mind co-workers from HPE ganging up together to cover up any project failures or mishaps with career advancement in mind. Nothing against the fresh blood into grommet - you guys/gals are doing a great job, but just i don't think this works long term for an OSS project to have work/career bias in the decision making process.

**Comment by juanlanus** [comment]
**Grid integration:**
Two or 3 years ago I worked for a huge NY bank in a team chartered with the integration of a data grid to the tooling.  
The chosen one was [SlickGrid](http://slickgrid.net/). 
What I want to say is that it was an 8 people team work for several months. That is, do not underestimate the effort to bring such controls in.  
On the other hand, for enterprise applications, it'a a must to provide access to a solid grid. 
But many enterprises prefer to purchase products provided by another enterprise, thinking that they will be somebody to talk with (for example most are using Windows). 
What I want to say is that onboarding open-source grids galore might be a hard task perhaps not that valuable. And also that there is a need for a really strong grid. 
Perhaps somebody at the enterprise level could ask privately to other enterprise Grommet users what are they using?

**Comment by atanasster** [comment]
@juanlanus i already added grommet support for the devexpress react grid control and am not underestimating such task - took me about 2-3 weeks, mostly because grommet table components are not really extensible in that sense. If they are, it would take maybe a week, and another week to write an article/docs

http://grommet-nextjs.herokuapp.com/templates/dx-grid
https://github.com/atanasster/dx-react-grid-grommet

**Comment by juanlanus** [comment]
**DraftJS**:
Onboarding this WYSIWYG text editor fy facebook could be useful. 
It seems that it's quickly becoming the defacto standard.

**Comment by juanlanus** [comment]
**RTL:**
Multinational enterprises will require Grommet to be committed to RTL text.

**Comment by brodybits** [comment]
-100 for a switch to TypeScript

I think the TypeScript language has additional complexity which offers little if any additional benefits for end users of web applications. It would introduce an additional barrier of entry for new contributions that I think is not really needed or wanted. I think it would be much better to use JSDoc comments to add static type checking to the existing JavaScript, as I proposed in #3353.

I think there are some excellent points by @ericelliott in the following articles:

* https://medium.com/javascript-scene/the-typescript-tax-132ff4cb175b
* https://medium.com/javascript-scene/you-might-not-need-typescript-or-static-types-aa7cb670a77b
* https://medium.com/javascript-scene/the-shocking-secret-about-static-types-514d39bf30a3
* https://medium.com/javascript-scene/angular-2-vs-react-the-ultimate-dance-off-60e7dfbc379c

see also this discussion: prettier/prettier#6286

My apologies for giving this feedback now and not earlier.

**Comment by atanasster** [comment]
@brodybits - while i disagree with your take, typescript is a decision that was already made 1+ years ago when v2 of grommet shipped the .d.ts typings. The problem is that it was amateurishly designed and wrongly implemented at the time, but has to be maintained today and in the future

- the initial .d.ts typings were generated by react-desc out of prop-types type of typings. Its an impossible task as i pointed(pointlessly) from the get go - typescript typings are much richer than prop-types.
- at a later stage grommet v2 dropped the very buggy react-desc d.ts typings and started writing manually the .d.ts files. While this contradicts some basic software rules (single source of truth) as you have to maintain two sets of types, It is an approach that works with some of the most popular es projects like react, react-dom. But fails with projects even much more popular than grommet (webpack, webpack-sources) - it simply takes too much effort to maintain a separate set of typings. Currently the v2 .d.ts are still incomplete and buggy and it takes almost a full time maintainer to keep fixing them. A contributor will also have to submit prop changes both to the react-desc prop tables and the d.ts files.
- the proposal here is simply to take the correct approach - write grommet in typescript and then generate from that all other /es, cjs/ required formats. Typescript is a superset of all those other formats, so this is the only way to reduce the maintenance work. As an added benefit, it will also help catch many bugs in v2 internals, but thats another discussion:)

As a footnote, i have a special place for typescript in my heart as it was architected by Anders Hejlsberg, the architect of Turbo Pascal and Delphi languages that I used for 20+ years :)

**Comment by sofiageo** [comment]
I have not much to add other than Typescript is great. I don't think it adds burden to developers,  I've been using it less than a week and it was really straight forward for someone that already knows some JS.

One other thing that comes to mind is maybe rewrite or, start writing components as functional components. I believe this is the way react team is adding new components to their project, and are probably easier to work with. 

p.s I could probably also help with refactoring to typescript, if required.

**Comment by srallen** [comment]
1. Highly recommend Grommet look into migrating the Markdown component from markdown-to-jsx to remark (plus remark2react) and supporting a way to incorporate additional remark plugins. We had to build our own Markdown component because markdown-to-jsx only supports Github-flavored markdown and we have some custom markdown that we have to support in our app.
2. Continued high priority for accessibility and theming. Related to that, infinite scroll is not considered to be accessible by my group (not that it's impossible, but it's really hard to do compared to an accessible pagination solution with a paginator component with appropriate toggleable sorts and filters). I would vote to move infinite scroll into an extras package or to drop it. 
 3. About charts, I would vote for only having very low level primitive components rather that an all-in-one solution that makes a lot of assumptions on how charts are being used. Because of the assumptions a lot of charting libraries make (i.e. plot.ly's react implementation and others), we could not use them and had to roll with d3.js. Now we're migrating to [vx](https://github.com/hshoff/vx) because this library is just low level primitive React components plus d3 for utility functions. We can build our custom requirements but save time and keep the DOM being handled by React now. If Grommet does support charts, I'd highly recommend taking a look at the approach vx is taking or just contribute to that project 😉 
4. +1 to making sure RTL languages work with Grommet

**Comment by atanasster** [comment]
@srallen +1 for remark, have you looked at https://github.com/mdx-js/mdx it uses remark-mdx under the hood. But then not sure a markdown component is needed at all unless the elements are styled with grommet components by default.

**Comment by srallen** [comment]
@atanasster yes, I have. Supporting jsx directly in the markdown wasn't a requirement for us, but rather supporting custom markdown syntax already in use by our users, so we were able to accomplish this directly with remark, remark2react, some additional existing remark plugins, and a custom plugin. I think mdx could be a good pick for Grommet though.

**Comment by nomadus** [comment]
> **DraftJS**:
> Onboarding this WYSIWYG text editor fy facebook could be useful.
> It seems that it's quickly becoming the defacto standard.

@juanlanus There was an attempt to add a text editor within a grommet-cms project. Per comment, it is outside of grommet core. https://github.com/grommet/grommet/issues/1196

**Comment by atanasster** [comment]
@nomadus thanks for the reference, fascinating to see almost 3 years later the amount of wasted work caused by v2. Searching the grommet github and npm is like a wasteland of dead grommet projects. God bless HPE for footing the bill...
I had a fully working draftjs + grommet v1 editor that i had to trash and we could team up to provide a working v2 solution if i see some seriously positive steps with v2.

**Comment by atanasster** [comment]
Going down memory lane, my first PR contributing the draftjs editor icons to grommmet-icons 2 years ago https://github.com/grommet/grommet-icons/pull/38

**Comment by dustykeyboard** [comment]
Looking forward to improved types. Theme types/validation is something that would surely help

**Comment by ShimiSun** [comment]
I shared a quick update on slack and I'll share it here as well
>...The discussions were very interesting and that mostly thanks to the fantastic feedback and comments you guys shared on the roadmap, we were happy to consider it all and had a lot to work with.
I saw a couple of more comments that came over the weekend, and we would like to have a couple of more internal discussions with all team (some currently traveling) before we can share our draft, hopefully, it will be before the end of the month or the beginning of the following month. Hope that works.

One comment to @brodybits, there is a phrase I like
*'If two people think the same thing, one of them isn't necessary.'*
So no worries about being controversial https://github.com/grommet/grommet/issues/3353#issuecomment-533357987
our decisions are more educated if we hear all the voices in the room, so thanks for speaking up.

**Comment by ericelliott** [comment]
Just popping in to say hello! 👋 

And to add: I love TypeScript. But it does have a high maintenance burden, particularly for large projects with many contributors with many different skill levels.

In my articles, I'm not trying to convince people not to use TypeScript. Here are some verbatim quotes that people seem to overlook because people like things to be simple black-and-white answers rather than the messy, nuanced grey that is the real world:

> "If you haven’t experienced Microsoft’s Code IDE paired with TypeScript, it’s worth kicking the tires. You might fall in love."

> "Many developers love using it, and there are many aspects of the TypeScript developer experience I genuinely love. But all of this comes with a cost."

> "TypeScript supports interfaces, which are reusable (as opposed to inline) typings that you can apply in various ways to annotate APIs and function signatures. A single interface can have many implementations. Interfaces are one of the best features of TypeScript, and I wish this feature was built into JavaScript."

> "The best news: If you use one of the well supported editors (such as Atom or Visual Studio Code), TypeScript’s editor plugins still provide the best IDE developer experience in the JavaScript ecosystem, in my opinion. Other plugin developers should try them out and take notes on how they can improve."

> "Keep in mind: I’m not arguing against using TypeScript. I’m arguing for people to consider the costs and benefits and make a rational, informed decision that’s right for you and your team. Some products require stricter quality control, and it may be worth the extra cost to eliminate 18 more bugs out of a thousand. For example, if your code powers critical parts of the self-driving system for a Tesla, I hope you’re using static types along with all the other quality measures, because the cost of bugs is much higher. Each team should conduct their own ROI analysis and make the decision that is right for them."

If TypeScript is the right decision for you and your team, there is *nothing wrong with that at all.*

You do what works for you.

Personally, I find that gradual typing using JSDoc and .d.ts files works better for me, and I wouldn't use the TypeScript language again on a large scale app if I were building from scratch, but it's not wrong to use TypeScript, and for many people (like the people at Google who get great benefits maintaining their monorepo), TypeScript may be exactly the right choice.

You probably will find a few bugs converting to TypeScript, particularly if your unit test coverage is less than perfect. You'll definitely find lots of other cool benefits, as your intellisense seems to grow more and more intelligent, and features like jump to definition start to work a bit better -- especially if you're not already using a type inference engine like Tern.js (even if I'm not using TypeScript on a project, I'm still getting type hints using type inference).

You may also find that as more libraries adopt TypeScript over time (this is the trend), the maintenance burden of using TypeScript will continue to go down over time, because you'll spend less time backfilling missing typings.

You'll probably also find that while that burden is going down gradually, it's still there. You'll also find TypeScript trains you not to do some things that are perfectly valid idiomatic JavaScript because TypeScript doesn't recognize it or know what to do with it. If you're a functional programmer with a high level of compositional thinking, you'll find yourself [writing extra pointed function wrappers around point-free compositions just to drop TypeScript a clue](https://twitter.com/JS_Cheerleader/status/1188909848794034176) because it's still not particularly good at data flow analysis.

But these are all tradeoffs that you may find to be worth the cost. Because there really is a lot to love about TypeScript.

Please don't think I hate it just because I've repeatedly warned developers that it has a cost. Lots of great things have costs.

Yes, it does have a cost. It also has benefits. It sounds like you have made a rational choice that makes sense to you for your project.  IMO, that's exactly what everyone should be doing.

**Comment by anuraghpe** [comment]
Consider adopting [Styled System](https://styled-system.com/) as the standard props based styling mechanism.

**Comment by anuraghpe** [comment]
Some problems with current implementation:

- `Grid` is limited. It only supports a very small set of features as compared to native CSS Grids.
- `gap` attribute is confusing. There is a CSS `gap` property, and the implementation seems to emulate this. But, it adds an extra `div` with a fixed height or width instead. These elements are unnecessary and have the potential to cause issues with layout.
- `Button` doesn't work as expected. When you add an `icon` to the button, you can't have plain text or any React Node as `children`. 
- Most components have limited customizability. They only support a limited set of CSS attributes. eg:  Can't adjust `font-size` on a `TableCell`. To work around these, you need to nest `Text` inside a `TableCell` which is unnecessary.
- Some attributes like `size` have different meaning for different components. eg: `Text` vs any other component.
- Responsiveness is very limited. The only component available is `ResponsiveContext`. It forces you to write javascript for something that is better supported by media queries.

These are just some of my observations. With my limited experience in using Grommet, I found there are a lot of surprises and unexpected behavior. Coming from a user of [ChakraUI](https://chakra-ui.com/). The library feels inconsistent and limited, even tough Grommet has a lot more components in comparison.

Styled System addresses a lot of these issues, specially related to CSS support: 

- It introduces a consistent set of attributes on **every** component to customize and modify them to your heart's content.
- Introduces a syntax to make every attribute responsive by using [arrays or objects](https://styled-system.com/guides/array-props). 
- The theme [specification](https://styled-system.com/theme-specification) makes the theme portable. I think a bigger advantage is that we get access to a lot of already existing themes.
- I'm sure there are a lot more issues described elsewhere and in earlier comments that can be addressed by Styled System. 
- A lot more information on it's [website](https://styled-system.com/).

I propose we use a base like [Theme UI](https://theme-ui.com/) or [Rebass](https://rebassjs.org/), or as a final resort, for complete control, build our own base based on styled-system and rewrite the whole library on this.
